certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,800 @@
+
+    require to_mathint(asset.balanceOf(vault)) == 2 * principal;
+ 
+    require receivers.length == allocations.length && allocations.length == 1;
+    require receivers[0] == bob && allocations[0] == principal;
+
+    uint256 shares = vault.deposit(e, principal, currentContract);
+    asset.approve(e, currentContract, shares);
+
+    // Capture the initial state
+    uint256 initialNextStreamId = nextStreamId();
+    uint256 initialBobShares = receiverTotalShares(bob);
+    uint256 initialBobPrincipal = receiverTotalPrincipal(bob);
+
+    // Call the depositAndOpenMultipleUsingPermit function
+    uint256[] streamIds = depositAndOpenMultipleUsingPermit(
+        e,
+        principal,
+        receivers,
+        allocations,
+        0,
+        deadline,
+        v,
+        r,
+        s
+    );
+
+    require ownerOf(streamIds[0]) == dave;
+
+    // Postconditions
+    assert streamIds.length == 1 &&
+     streamIds[0] == initialNextStreamId &&
+     to_mathint(nextStreamId()) == initialNextStreamId + 1;
+    assert to_mathint(receiverTotalPrincipal(bob)) == initialBobPrincipal + principal &&
+     receiverPrincipal(bob, streamIds[0]) == principal && vault.balanceOf(currentContract) == shares;
+}
+
+
+/**
+ * @Rule Integrity property for the `topUp` function
+ * @Category High
+ * @Description Ensures that the `topUp` function adds the specified shares to the existing yield stream and update the contract state accordingly, if the caller is the owner of the stream.
+ */
+
+rule integrity_of_topUp(uint256 _streamId, uint256 _shares) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 principal;
+
+    // Preconditions
+    require _shares > 0;
+    require ownerOf(_streamId) == streamer;
+    require receiver != 0;
+
+    // Capture the initial state
+    uint256 initialTotalShares = receiverTotalShares(receiver);
+    uint256 initialTotalPrincipal = receiverTotalPrincipal(receiver);
+    uint256 initialPrincipal = receiverPrincipal(receiver, _streamId);
+
+    // Call the `topUp` function
+    principal = topUp(e, _streamId, _shares);
+
+    // Postconditions
+    assert to_mathint(receiverTotalShares(receiver)) == initialTotalShares + _shares;
+    assert to_mathint(receiverTotalPrincipal(receiver)) == initialTotalPrincipal + principal;
+    assert to_mathint(receiverPrincipal(receiver, _streamId)) == initialPrincipal + principal;
+}
+
+/**
+ * @Rule Integrity property for the `topUpPermit` function
+ * @Category High
+ * @Description Ensures that the `topUpPermit` function correctly adds additional shares to an existing yield stream and updates the contract state.
+ */
+
+rule integrity_of_topUpPermit(uint256 _streamId, uint256 _shares, uint256 deadline, uint8 v, bytes32 r, bytes32 s) {
+    env e;
+    address streamer = e.msg.sender;
+    address receiver = streamIdToReceiver(_streamId);
+    uint256 principal;
```

Comment:
It would be easier to read if moving the declaration to the assignment below

Reply:
Updated